src/core/build_env.go

```diff
@@ -220,74 +184,115 @@ func TestEnvironment(state *BuildState, target *BuildTarget, testDir string) Bui
-		} else {
-			env = append(env, "DATA="+strings.Join(runtimeDataPaths(state.Graph, target.AllData()), " "))
-		}
-	}
-	if target.namedData != nil {
-		for name, data := range target.namedData {
-			var paths []string
-			if inTmpDir {
-				paths = target.SourcePaths(state.Graph, data)
-			} else {
-				paths = runtimeDataPaths(state.Graph, data)
-			}
-			env = append(env, "DATA_"+strings.ToUpper(name)+"="+strings.Join(paths, " "))
-		}
-	}
 	return withUserProvidedEnv(target, env)
 }
 
 // ExecEnvironment creates the environment variables for a `plz exec`.
 func ExecEnvironment(state *BuildState, target *BuildTarget, execDir string) BuildEnv {
-	env := TargetEnvironment(state, target)
-
-	env = append(env, "TMP_DIR="+filepath.Join(RepoRoot, execDir))
+	env := append(RuntimeEnvironment(state, target, true, true),
+		"TMP_DIR="+execDir,
+		"TMPDIR="+execDir,
+		"HOME="+execDir,
+		// This is used by programs that use display terminals for correct handling
+		// of input and output in the terminal where the program is run.
+		"TERM="+os.Getenv("TERM"),
+	)
 
 	outEnv := target.Outputs()
+	// OUTS/OUT environment variables being always set is for backwards-compatibility.
+	// Ideally, if the target is a test these variables shouldn't be set.
 	env = append(env, "OUTS="+strings.Join(outEnv, " "))
 	if len(outEnv) == 1 {
-		env = append(env, "OUT="+outEnv[0])
+		env = append(env, "OUT="+resolveOut(outEnv[0], ".", target.Sandbox))
+		if target.IsTest() {
+			env = append(env, "TEST="+resolveOut(outEnv[0], ".", target.Test.Sandbox))
+		}
 	}
 
 	return withUserProvidedEnv(target, env)
 }
 
+// RuntimeEnvironment is the base environment for runtime-based environments.
+// Tools and data env variables are made available.
+func RuntimeEnvironment(state *BuildState, target *BuildTarget, abs, inTmpDir bool) BuildEnv {
+	env := TargetEnvironment(state, target)
+
+	// Data
+	env = append(env, dataEnv(state, target.AllData(), target.NamedData, "", inTmpDir)...)
+
+	if target.IsTest() {
+		// Test tools
+		env = append(env, toolsEnv(state, target.AllTestTools(), target.NamedTestTools(), "", abs)...)
+	}
+
+	if target.Debug != nil {
+		prefix := "DEBUG_"
+		// Debug data
+		env = append(env, dataEnv(state, target.AllDebugData(), target.DebugNamedData(), prefix, inTmpDir)...)
+		// Debug tools
+		env = append(env, toolsEnv(state, target.AllDebugTools(), target.Debug.namedTools, prefix, abs)...)
+	}
+
+	return env
+}
+
+// Handles resolution of OUT files
+func resolveOut(out string, dir string, sandbox bool) string {
+	// Bit of a hack; ideally we would be unaware of the sandbox here.
+	if sandbox && runtime.GOOS == "linux" && !strings.HasPrefix(RepoRoot, "/tmp/") && dir != "." {
+		return path.Join(SandboxDir, out)
+	}
+	return path.Join(dir, out)
+}
+
+// Creates tool-related env variables
+func toolsEnv(state *BuildState, allTools []BuildInput, namedTools map[string][]BuildInput, prefix string, abs bool) BuildEnv {
```

Comment:
This refactored tool-related env variables into this function. It now guarantees that they are constructed consistently across different environments. There used to be some mismatches before. This follows the pattern used by `$SRCS`